### DISCUSSION TITLE
project_tbrom_on_mesh can't project result properly

### DISCUSSION THREAD

**Discussion by willywah** [ansys/pytwin#167] [General]
Hi, I want to know whether this is a bug or an error in my code. I used the `project_tbrom_on_mesh` API to project my results from the twin file onto the FEA mesh from the RST file. I have checked the TBROM data inside the twin file many times and found no errors (I have attached comments for exporting the TBROM data and plotting the results).

I used a simply supported beam model. Here is an example of the deformation result:
![image](https://github.com/ansys/pytwin/assets/95516289/c12282c4-c5c8-4745-9082-5bde5fc4e174)

When I use `project_tbrom_on_mesh`, the result does not represent the actual result, and the interpolation fails (if any).
![image](https://github.com/ansys/pytwin/assets/95516289/f7b0324d-845b-45be-b27b-5ad2ff9f8265)

For both interpolation settings (True or False), the result of the projection is the same. This error occurs for both deformation and stress.

Can anyone help me?

[project_tbrom_on_fea.zip](https://github.com/user-attachments/files/16061663/project_tbrom_on_fea.zip)
[project_tbrom_on_fea.zip](https://github.com/user-attachments/files/16066296/project_tbrom_on_fea.zip) (updated file)

**Comment by chrpetre** [comment]
hello @WillyW28 in order to investigate this issue further, could you please share the rst file too ?

**Reply by willywah** [reply]
hi @chrpetre sure, I already updated

**Comment by chrpetre** [comment]
hi @WillyW28 ,

it seems the rst file has point coordinates defined in mm (I guess) while Twin/ROM has its points defined in m. Can you try the following : 

whole_mesh = dpf.operators.mesh.mesh_provider(streams_container=streams).eval()
target_mesh = whole_mesh.grid
**target_mesh.points /= 1000.0**

with that modification, I get something as follows :

![image](https://github.com/ansys/pytwin/assets/108675940/f9522e73-1fc9-48dd-87c0-81c571c8527b)

let me know if that solves your problem

**Reply by willywah** [reply]
Hi @chrpetre, 

Thanks for assisting a small problem here. The `project_tbrom_on_mesh` API works properly for the deformation case, and the results make sense as well. However, for the stress case, I have already validated with the Static Rom Builder provided by Twin Builder. The maximum stress value (I assumed Von Mises Stress) from the `project_tbrom_on_mesh` plot is always less than the reference ROM (von Mises stress). In this case, the maximum stress value is around half of the reference ROM.

Here is the results with the input `"Force_Magnitude": 20000
`![image](https://github.com/ansys/pytwin/assets/95516289/0c0a803a-c342-4df2-b195-67fec6c764b5)
![image](https://github.com/ansys/pytwin/assets/95516289/f93148a4-c3d2-41ea-b358-66cd1cbf4b7b)

From both images above, we notice that the stress distribution is also different in the neutral axis of the beam.

I also added some lines for validation, and it gives the reference ROM value. 

```
outfield = twin_model.generate_snapshot(rom_name, on_disk=False)
def von_mises_stress(stress_components):
    sigma_x, sigma_y, sigma_z, tau_xy, tau_yz, tau_zx = stress_components
    von_mises = np.sqrt(0.5 * ((sigma_x - sigma_y)**2 + (sigma_y - sigma_z)**2 + (sigma_z - sigma_x)**2 +
                               6 * (tau_xy**2 + tau_yz**2 + tau_zx**2)))
    return von_mises

# Compute von Mises stress for each set of stress components
von_mises_stresses = np.apply_along_axis(von_mises_stress, 1, outfield)

# Find the maximum von Mises stress value
max_von_mises_stress = np.max(von_mises_stresses)

print("Maximum von Mises stress:", max_von_mises_stress)
 
```

**Comment by chrpetre** [comment]
hi @WillyW28, I see the projection is done with interpolation since the ROM number of points (1723) is not the same as the number of points in the rst file (1521).
Can you explain the differences between the 2 meshes (the one used to generated training data/ROM, and the one used for projection) ? Is there any chance to use the same mesh for both parts ?

**Reply by willywah** [reply]
hello @chrpetre, the mesh model for projection I used is actually obtained from the Workbench file (.../dp0/SYS/MECH/file.rst). However, I used the StaticROM Pre add-on in Workbench, and it provided snapshot.bin files. Then, I used the Static ROM Builder in Twin Builder to generate a ROM from the source files (DOE, bin, etc.). I think the difference is that one is from the rst file, and the other one is from the snapshot.bin file.

Do you have any advice on this?

**Reply by willywah** [reply]
Hi, I think I know the problem here. I was generating a single beam body and creating two additional named selections on the edges (the upper edge and the lower edge of the body). In the StaticROM Pre add-on, I exported these selections separately as outputs. Somehow, the .rst file cannot recognize the named selections, maybe because it only contains a single body. Meanwhile, the ROM detects additional points as I create named selections on the edges, even though there are already points in the same positions.

this is the config files generated by StaticROM Pre add-on as references:
[settings.json](https://github.com/user-attachments/files/16095744/settings.json)
[settings.json](https://github.com/user-attachments/files/16095746/settings.json) (stress)
[settings.json](https://github.com/user-attachments/files/16095750/settings.json) (deformation)
